File: docs/chapters/20-distributed-systems-fundamentals/java/RetryWithJitter.java
// Scenario: Checkout calls payment provider; transient 503s need bounded retries
// Demonstrates: exponential backoff + full jitter; idempotency-key on every attempt
// Trade-off: more latency on failure vs retry storm without jitter

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.concurrent.ThreadLocalRandom;

final class RetryWithJitter {
  private final HttpClient http = HttpClient.newHttpClient();
  private final String payBase;

  RetryWithJitter(String payBase) { this.payBase = payBase; }

  boolean charge(String idempotencyKey, String orderId, Duration totalBudget) throws InterruptedException {
    long deadline = System.nanoTime() + totalBudget.toNanos();
    int attempt = 0;
    while (System.nanoTime() < deadline && attempt < 4) {
      HttpRequest req = HttpRequest.newBuilder()
          .uri(java.net.URI.create(payBase + "/charge"))
          .timeout(Duration.ofMillis(400))
          .header("Idempotency-Key", idempotencyKey)
          .POST(HttpRequest.BodyPublishers.ofString("{\"orderId\":\"" + orderId + "\"}"))
          .build();
      try {
        HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
        if (resp.statusCode() == 200) return true;
        if (resp.statusCode() < 500) return false;
      } catch (Exception ignored) { /* retryable */ }
      sleepJitter(attempt++);
    }
    return false;
  }

  private static void sleepJitter(int attempt) throws InterruptedException {
    long cap = Math.min(2000, 100L << attempt);
    Thread.sleep(ThreadLocalRandom.current().nextLong(cap));
  }
}

File: docs/chapters/20-distributed-systems-fundamentals/go/retry_with_jitter.go
// Scenario: Checkout calls payment provider; transient errors need bounded retries
// Demonstrates: exponential cap + full jitter; same idempotency-key each attempt
// Trade-off: explicit error return vs Java exceptions — same retry policy

package main

import (
	"context"
	"math/rand"
	"net/http"
	"strings"
	"time"
)

func charge(ctx context.Context, client *http.Client, baseURL, idempotencyKey, orderID string) (bool, error) {
	deadline, ok := ctx.Deadline()
	if !ok {
		deadline = time.Now().Add(3 * time.Second)
	}
	for attempt := 0; attempt < 4 && time.Now().Before(deadline); attempt++ {
		reqCtx, cancel := context.WithTimeout(ctx, 400*time.Millisecond)
		req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, baseURL+"/charge",
			strings.NewReader(`{"orderId":"`+orderID+`"}`))
		if err != nil {
			cancel()
			return false, err
		}
		req.Header.Set("Idempotency-Key", idempotencyKey)
		resp, err := client.Do(req)
		cancel()
		if err == nil {
			resp.Body.Close()
			if resp.StatusCode == http.StatusOK {
				return true, nil
			}
			if resp.StatusCode < 500 {
				return false, nil
			}
		}
		jitterSleep(attempt)
	}
	return false, context.DeadlineExceeded
}

func jitterSleep(attempt int) {
	cap := min(2000, 100<<attempt)
	time.Sleep(time.Duration(rand.Int63n(int64(cap))) * time.Millisecond)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
